File: NutriSmart/app/src/main/java/com/example/nutrismart/ui/screens/login/LoginViewModel.kt
package com.example.nutrismart.ui.screens.login

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.nutrismart.data.UserSession
import com.example.nutrismart.data.model.AuthRequest
import com.example.nutrismart.data.model.GoogleLoginRequest
import com.example.nutrismart.data.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class LoginViewModel @Inject constructor(
    private val authRepository: AuthRepository
) : ViewModel() {

    var email by mutableStateOf("")
    var password by mutableStateOf("")
    var isLoading by mutableStateOf(false)
    var errorMessage by mutableStateOf<String?>(null)
    var loginSuccess by mutableStateOf(false)
    var isNewUser by mutableStateOf(false)
    var isVerified by mutableStateOf(true)

    fun onEmailChange(newEmail: String) {
        email = newEmail
    }

    fun onPasswordChange(newPassword: String) {
        password = newPassword
    }

    fun resetNavigation() {
        loginSuccess = false
    }

    fun login() {
        isLoading = true
        errorMessage = null

        viewModelScope.launch {
            try {
                val request = AuthRequest(email, password)
                val response = authRepository.login(request)

                if (response.isSuccessful) {
                    response.body()?.let { authData ->
                        UserSession.currentUserId = authData.userId
                        UserSession.token = authData.token

                        authRepository.saveAuthData(authData)
                        authRepository.saveIsGoogleUser(false)

                        isNewUser = !authData.isProfileComplete
                        isVerified = authData.isVerified
                        loginSuccess = true
                    }
                } else {
                    errorMessage = "Login failed: Check data."
                }
            } catch (e: Exception) {
                errorMessage = "Connection error: ${e.message}"
                e.printStackTrace()
            } finally {
                isLoading = false
            }
        }
    }

    fun loginWithGoogle(idToken: String) {
        isLoading = true
        errorMessage = null

        viewModelScope.launch {
            try {
                val request = GoogleLoginRequest(idToken)
                val response = authRepository.googleLogin(request)

                if (response.isSuccessful) {
                    response.body()?.let { authData ->
                        UserSession.currentUserId = authData.userId
                        UserSession.token = authData.token

                        authRepository.saveAuthData(authData)
                        authRepository.saveIsGoogleUser(true)

                        isNewUser = !authData.isProfileComplete
                        isVerified = authData.isVerified
                        loginSuccess = true
                    }
                } else {
                    errorMessage = "Google login failed: ${response.code()}"
                }
            } catch (e: Exception) {
                errorMessage = "Error connecting to server: ${e.message}"
            } finally {
                isLoading = false
            }
        }
    }
}

File: NutriSmartIOS/ui/screens/login/LoginViewModel.swift
//
//  LoginViewModel.swift
//  NutriSmartIOS
//

import Foundation
import SwiftUI
import Combine

@Observable
@MainActor
class LoginViewModel {
    var email = ""
    var password = ""
    var isLoading = false
    var errorMessage: String? = nil
    
    var loginSuccess = false
    var isNewUser = false
    var isVerified = true
    
    private let authRepository = AuthRepository.shared
    
    func resetNavigation() {
        loginSuccess = false
    }
    
    func onEmailChange(_ newEmail: String) {
        email = newEmail
    }
    
    func onPasswordChange(_ newPassword: String) {
        password = newPassword
    }
    
    func login() {
            isLoading = true
            errorMessage = nil
            
            Task {
                do {
                    let request = AuthRequest(email: email, password: password)
                    let response = try await authRepository.login(request: request)
                    
                    UserSession.shared.currentUserId = response.userId
                    UserSession.shared.token = response.token
                    
                    authRepository.saveAuthData(authData: response)
                    SessionManager.shared.saveIsGoogleUser(false)
                    
                    isNewUser = !response.isProfileComplete
                    isVerified = response.isVerified
                    loginSuccess = true
                    
                } catch {
                    self.errorMessage = "Login failed: \(error.localizedDescription)"
                }
                isLoading = false
            }
        }
        
        func loginWithGoogle(idToken: String) {
            isLoading = true
            errorMessage = nil
            
            Task {
                do {
                    let request = GoogleLoginRequest(token: idToken)
                    let response = try await authRepository.googleLogin(request: request)
                    
                    UserSession.shared.currentUserId = response.userId
                    UserSession.shared.token = response.token
                    
                    authRepository.saveAuthData(authData: response)
                    SessionManager.shared.saveIsGoogleUser(true)
                    
                    isNewUser = !response.isProfileComplete
                    isVerified = response.isVerified
                    loginSuccess = true
                    
                } catch {
                    self.errorMessage = "Google login failed: \(error.localizedDescription)"
                }
                isLoading = false
            }
        }
}
